src/devices/src/virtio/block/device.rs

```diff
@@ -0,0 +1,263 @@
+            return Err(Error::BadFeatures(self.virtio_cfg.driver_features));
+        }
+
+        // Set the appropriate queue configuration flag if the `EVENT_IDX` features has been
+        // negotiated.
+        if self.virtio_cfg.driver_features & (1 << VIRTIO_F_RING_EVENT_IDX) != 0 {
+            self.virtio_cfg.queues[0].set_event_idx(true);
+        }
+
+        let ioeventfd = EventFd::new(EFD_NONBLOCK).map_err(Error::EventFd)?;
+
+        // Register the queue event fd.
+        self.vm_fd
+            .register_ioevent(
+                &ioeventfd,
+                &IoEventAddress::Mmio(
+                    self.mmio_cfg.range.base().0 + VIRTIO_MMIO_QUEUE_NOTIFY_OFFSET,
+                ),
+                0u32,
+            )
+            .map_err(Error::RegisterIoevent)?;
+
+        // TODO: Some changes required here to add read-only support as well.
+        let file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open(&self.file_path)
+            .map_err(Error::OpenFile)?;
+
+        // TODO: Create the backend earlier (as part of `Block::new`)?
+        let disk =
+            StdIoBackend::new(file, self.virtio_cfg.driver_features).map_err(Error::Backend)?;
+
+        let driver_notify = SingleFdSignalQueue {
+            irqfd: self.irqfd.clone(),
+            interrupt_status: self.virtio_cfg.interrupt_status.clone(),
+        };
+
+        let inner = InOrderQueueHandler {
+            driver_notify,
+            queue: self.virtio_cfg.queues[0].clone(),
+            disk,
+        };
+
+        let handler = Arc::new(Mutex::new(QueueHandler { inner, ioeventfd }));
+
+        // Register the queue handler with the `EventManager`. We could record the `sub_id`
+        // (and/or keep a handler clone) for further interaction (i.e. to remove the subscriber at
+        // a later time, retrieve state, etc).
+        let _sub_id = self
+            .endpoint
+            .call_blocking(move |mgr| -> EvmgrResult<SubscriberId> {
+                Ok(mgr.add_subscriber(handler))
+            })
+            .map_err(Error::Endpoint)?;
+
+        self.virtio_cfg.device_activated = true;
+
+        Ok(())
+    }
+
+    fn reset(&mut self) -> Result<()> {
+        // Not implemented for now.
+        Ok(())
+    }
+}
+
+impl<M: GuestAddressSpace + Clone + Send + 'static> VirtioMmioDevice<M> for Block<M> {}
+
+impl<M: GuestAddressSpace + Clone + Send + 'static> MutDeviceMmio for Block<M> {
+    fn mmio_read(&mut self, _base: MmioAddress, offset: u64, data: &mut [u8]) {
+        self.read(offset, data);
+    }
+
+    fn mmio_write(&mut self, _base: MmioAddress, offset: u64, data: &[u8]) {
+        self.write(offset, data);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use vmm_sys_util::tempfile::TempFile;
+
+    use crate::virtio::tests::CommonArgsContext;
+
+    use super::*;
+
+    // Restricting this for now, because registering irqfds does not work on Arm without properly
```

Comment:
Besides the bus range check anything else looks very x86_64 specific at the moment. To make it less specific we would need to abstract the way we specify devices (cmdline on x86), and update the CommonArgsContext such that it does not also create the irqchip. It should instead receive just a reference to the VmFd and assume that the irqchip is already created so that we can reuse the code on arm as well.

Since it is not really reusable at the moment, I would just make this test x86_64 specific with `cfg(target_arch = x86_64)`.

Reply:
Hmm, this is not meant to be `x86_64` specific ... besides the current issue that prevents it from working properly on `aarch64`, what else looks x86 specific? Also, I'd keep the `ignore` so the test explicitly shows up as `ingored` as an additional signal/reminder that we need to fix it, but we can drop this if it's too verbose.

Reply:
It is specific because:
- the mock is calling into vm.create_irq_chip (I left a suggestion where it is introduced about how this could become less specific)
- it is checking the side effect introduced by the kernel commandline update which will not be present on aarch64.

In other words, the only check that we can do on both platform from the entire test is this one:
```
let bus_range = ctx.mmio_mgr.mmio_device(range.base()).unwrap().0;
assert_eq!(bus_range.base(), range.base());
assert_eq!(bus_range.size(), range.size());
```

Reply:
Cool, thanks for explaining! So, to add a bit of nuance, the current logic appears to be mostly for `x86` because the Arm specific stuff is not there yet to test, but when both are available, we should have platform specific-checks as well in here. I like that approach to avoid duplication of the common things, but we'll have a better perspective when Arm support starts getting added.